R/deps_installation_proposal.R

```diff
@@ -9,14 +9,25 @@
 #'
 #' @section strategies:
 #' Currently implemented strategies:
-#' * `max` - use the greatest version of dependent packages
-#' * `release` - use released version of dependent packages - use CRAN if possible else if GitHub release
-#' is available then use it else fail.
-#' * `min` - use the lowest version of dependent packages incorporating minimal version specification in
-#' `"Imports"` and `"Suggests"`. If no version is specified then the minimal available
-#' version is assumed. See [get_ref_min] for details how the minimal version is determined.
+#' * `max` - use the greatest version of dependent packages. Please note that using development version is not
+#' guaranteed to be stable.
+#' See [get_ref_max] for details.
+#' * `release` - use the released version of dependent packages. It will try use CRAN if possible else if
+#' GitHub release is available then use it else fail.
+#' See [get_ref_release] for details.
+#' * `min_cohort` - find maximum date of directly dependent packages release dates and use that as PPM snapshot date
+#' for dependency resolve.
+#' * `min_isolated` - for each direct dependency: find its release date and use it as PPM snapshot for resolving itself.
+#' Next, combine all the individual resolutions and resolve it altogether again.
 #'
-#' Any modification is done for direct (!) dependencies. Indirect ones are installed as usual.
+#' Both "min" strategies relies on PPM snapshot in order to limit the versions of indirect dependencies so that
+#' dependency resolution ends with a package released no earlier than any of its dependency.
+#' However, that's not always true for `min_isolated` strategy - done on purpose.
+#'
+#' Please note that only `min_cohort` and `min_isolated` strategies are "stable". The rest are basing on dynamic
+#' references therefore it results might be different without changes in tested package.
+#' The most straightforward example is `max` strategy in which the environment will be different after any push of
```

Comment:
I still think its a  mistake to do an install for non-version-bumping commits in main.

The reason is that when you get a positive result and then tell peoiple "yeah, this works for vx.y.z of my dependency", that isnt' actually true, because you don't have version x.y.z, you have some weird intermediary state that isn't a real version of a package at all.

Reply:
Technically you are right because package version is imperfect state of code identifier. It would be more correct to use md5 sum or check its source. This can give you unique identifier - as opposed to version. The gist is that this is not well known by users but here all the relevant info is accessible (and then debug-print in the action).

But you have one more interesting comment:

> you have some weird intermediary state that isn't a real version of a package at all

Well that's also true but this is somewhat expected, i.e. to take the development version (even if by definition it could be broken) and try to check your code using it. This needs to be a "soft" / "non obligatory" type of check that should not block merges or anything like that. I hope I made it very clear in docs. The overall goal for this strategy is not to make your code bulletproof but rather notify as soon as possible about upcoming (potentially breaking) changes.

Reply:
Well that sort of depends on what you think a "Development version of a package" is. Bioconductor, for example, has a formal development repository that only ever has distinctly named package versions in it, ever. GRAN is the same way. 

I think you need to sit down and decide which problem you're trying to solve, because they aren't the same and in fact they ultimately end up being incompatible.

Are you

a) building a tool for CI which tests whether the repo HEADs for multiple packages from SCM work together in an exact, "not real package versions" moment, _**or**_ 
b) building a tool which deals in package versions and informs package dependencies (the versioned dependencies referenced in the pkg name).

Currently, as far as I can tell, some parts of your tool are targeting goal (a), while others (and, I believe, the initial conception of the tool) target goal (b).

Both goals have value. Personally I think goal (b) has *more* value, because while the information it gives is less than that given in (a), that difference is marginal, and said information more correctly fits what users and CRAN maintainers care about anyway.

Reply:
You are right in your thinking. Historically, I started with b) using GH references only and then realized that a) can be also achieved with no additional input from user side. I rephrased the goal of "min deps testing" into "deps testing" and introduces a _strategy_ concept that help distinguish between those. The truth is that each of the _strategy_ answers different questions as you noticed and should be considered a little bit separate. The common ground is that it all relates to dependencies (actually env created based on deps) as well as I wanted to keep roughly the same input requirements for each of them (new field in DESC). The latter is not really true anymore but it's rather due to simplification (i.e. we don't need it) as opposed to adding new input requirement for a given strategy.
I really tried to explain and emphasise this difference in docs, readme, DESCRIPTION etc. Please let me know if you think it's not enough.